scripts/west_commands/packages.py

```diff
@@ -0,0 +1,138 @@
+# Copyright (c) 2024 Basalte bv
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import os
+import subprocess
+import sys
+import textwrap
+from itertools import chain
+from pathlib import Path
+
+from west.commands import WestCommand
+from zephyr_ext_common import ZEPHYR_BASE
+
+sys.path.append(os.fspath(Path(__file__).parent.parent))
+import zephyr_module
+
+
+def in_venv() -> bool:
+    return sys.prefix != sys.base_prefix
+
+
+class Packages(WestCommand):
+    def __init__(self):
+        super().__init__(
+            "packages",
+            "manage packages for Zephyr",
+            "List and Install packages for Zephyr and modules",
+            accepts_unknown_args=True,
+        )
+
+    def do_add_parser(self, parser_adder):
+        parser = parser_adder.add_parser(
+            self.name,
+            help=self.help,
+            description=self.description,
+            formatter_class=argparse.RawDescriptionHelpFormatter,
+            epilog=textwrap.dedent(
+                """
+            Listing packages:
+
+                Run 'west packages <manager>' to list packages for Zephyr and active modules
+                for a given package manager.
+                See 'west packages <manager> --help' for details.
+            """
+            ),
+        )
+
+        parser.add_argument(
+            "-p",
+            "--project",
+            action="append",
+            default=[],
+            dest="projects",
+            help="Project to run the 'packages' command for. "
+            "Use 'zephyr' if the 'packages' command should run for Zephyr itself. "
+            "Option can be passed multiple times. "
+            "If this option is not given, the 'packages' command will run for Zephyr "
+            "and all active modules.",
+        )
+
+        subparsers_gen = parser.add_subparsers(
+            metavar="<manager>",
+            dest="manager",
+            help="select a manager.",
+            required=True,
+        )
+
+        pip_parser = subparsers_gen.add_parser(
+            "pip",
+            help="manage pip packages",
+            formatter_class=argparse.RawDescriptionHelpFormatter,
+            epilog=textwrap.dedent(
+                """
+            Manage pip packages:
+
+                Run 'west packages pip' to print all pip packages or requirements
+                files needed by Zephyr and the active projects.
+
+                The output is compatible with the requirements file format itself.
+            """
+            ),
+        )
+
+        pip_parser.add_argument(
+            "--install",
+            action="store_true",
+            help="Install pip packages instead of listing them. "
+            "A single 'pip install' command is built and executed, passing all arguments "
+            "from the original 'west packages pip --install' command. For example pass "
+            "'--dry-run' to not actually install anything, but print what would be.",
```

Comment:
although i'm fine with keeping a `--project` / `-p` argument, then I still think we should separate native tool arguments, like `--dry-run` from the `west packages` arguments.

So that this is allowed:
```
$ west packages pip --install -- --dry-run
```

This will be consistent with other tools and also makes it clear what goes to the native tool and what is handled by the extension command itself.

Current result if trying above command is:
```
$  west packages pip --install -- --dry-run
ERROR: Invalid requirement: '--dry-run': Expected package name at the start of dependency specifier
    --dry-run
    ^

Traceback (most recent call last):
  File "/home/tora/zephyr-venv/lib/python3.10/site-packages/west/app/main.py", line 559, in run_command
    self.run_extension(args.command, argv)
  File "/home/tora/zephyr-venv/lib/python3.10/site-packages/west/app/main.py", line 716, in run_extension
    self.cmd.run(args, unknown, self.topdir, manifest=self.manifest,
  File "/home/tora/zephyr-venv/lib/python3.10/site-packages/west/commands.py", line 194, in run
    self.do_run(args, unknown)
  File "/projects/github/ncs/zephyr/scripts/west_commands/packages.py", line 99, in do_run
    return self.do_run_pip(args, unknown)
  File "/projects/github/ncs/zephyr/scripts/west_commands/packages.py", line 127, in do_run_pip
    subprocess.check_call(
  File "/usr/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/home/tora/zephyr-venv/bin/python', '-m', 'pip', 'install', '-r', PosixPath('/projects/github/ncs/zephyr/scripts/requirements.txt'), '--', '--dry-run']' returned non-zero exit status 1.

...
```

Reply:
Updated